pip/vcs/git.py

```diff
@@ -74,27 +74,34 @@ def export(self, location):
                 show_stdout=False, cwd=temp_dir.path
             )
 
-    def check_rev_options(self, rev, dest, rev_options):
-        """Check the revision options before checkout to compensate that tags
-        and branches may need origin/ as a prefix.
-        Returns the SHA1 of the branch or tag if found.
+    def get_ref_sha(self, dest, ref):
         """
-        revisions = self.get_short_refs(dest)
-
-        origin_rev = 'origin/%s' % rev
-        if origin_rev in revisions:
-            # remote branch
-            return [revisions[origin_rev]]
-        elif rev in revisions:
-            # a local tag or branch name
-            return [revisions[rev]]
-        else:
+        Return a commit hash for the reference if the given name is a ref.
+        Otherwise, return None.
+
+        Args:
+          dest: the repository directory.
+          ref: the ref name.
+        """
+        output = self.run_command(['show-ref', '--hash', ref], cwd=dest,
+                                  show_stdout=False, on_returncode='ignore')
+        # The git-show-ref command can return multiple references, so we
+        # return only the first.
+        for sha in output.strip().splitlines():
+            return sha
```

Comment:
I suppose the user would have to go mucking about in their installed clone to have any remote other than `origin`, so maybe that's not an edge case we should consider.

However, it occurred to me that the same problem exists with branches. I made an example repo https://github.com/di/git_ref_example which has four branches:
* `master` - prints "You are on branch `master`" when installed
* `blah` - prints "You are on branch `blah`" when installed
* `aaa/blah` - prints "You are on branch `aaa/blah`" when installed
* `zzz/blah` - prints "You are on branch `zzz/blah`" when installed

When I install the `blah` ref with `pip==9.0.1`, I get the `blah` branch as expected:
```
$ pip --version
pip 9.0.1 from /private/tmp/env/lib/python3.6/site-packages (python 3.6)
$ pip freeze
$ pip install git+https://github.com/di/git_ref_example@blah --verbose
...
    You are on branch `blah`
...
```

However if I do the same after installing the changes in this PR, I don't:
```
$ pip install git+https://github.com/cjerdonek/pip@simplify-git-ref-detection
$ pip --version
pip 10.0.0.dev0 from /private/tmp/env/lib/python3.6/site-packages (python 3.6)
$ pip freeze
$ pip install git+https://github.com/di/git_ref_example@blah --verbose
...
    You are on branch `aaa/blah`
...
```

This is because when you run `get_ref_sha` here, this is what you get:
```
$ git show-ref blah
59354f1a353783073ad80dfe619ccbe4ac38fd2b refs/heads/aaa/blah
f04115ccf5fb9e5ee5c89e15c0e58928a61719d9 refs/heads/blah
316b5d2bb9cd4e25486a3b22dcd006cf4f9b7836 refs/heads/zzz/blah
```

And since you're taking the first one, we get `aaa/blah`.